cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/common/plugin/JoinerBridge.java

```diff
@@ -48,8 +48,15 @@
   private final Map<String, List<String>> joinKeys;
   private final Map<String, List<JoinField>> stageFields;
   private Schema keySchema;
-
-  public JoinerBridge(BatchAutoJoiner autoJoiner, JoinDefinition joinDefinition) {
+  private Schema outputSchema;
+
+  public JoinerBridge(String stageName, BatchAutoJoiner autoJoiner, JoinDefinition joinDefinition) {
+    // if this is not an inner join and the output schema is not set,
+    // we have no way of determining what the output schema should be and need to error out.
+    if (joinDefinition.getOutputSchema() == null &&
+      joinDefinition.getStages().stream().anyMatch(s -> !s.isRequired())) {
+      throw new IllegalArgumentException("An output schema could not be generated for joiner stage " + stageName);
```

Comment:
Feel this should generate more meaningful error messages like in the comment to tell user what is the expected behavior to fix the problem